llvm-external-projects/iree-dialects/test/Dialect/iree_vector_ext/roundtrip.mlir

```diff
@@ -19,6 +19,54 @@ func.func @specify_layout(%lhs: memref<32x32xf16>) -> vector<32x32xf16> {
 // CHECK-SAME:         desiredLayout = #[[LAYOUT0]]
 // CHECK-SAME:         sourceLayout = #[[LAYOUT1]]
 
+// -----
+
+#layoutv2_1 = #iree_vector_ext.layoutv2<
+  subgroups_per_workgroup = [1, 1],
+  batches_per_subgroup = [2, 4],
+  duplicates_per_batch = [4, 1],
+  threads_per_duplicate = [4, 2],
+  elements_per_thread = [1, 4],
+
+  subgroup_order = [0, 1],
+  batch_order = [0, 1],
+  duplicate_order = [1, 0],
+  thread_order = [0, 1],
+  element_order = [0, 1]
```

Comment:
Reading through this, I'm not sure how this captures the same information as the current layout. Ignoring duplicates for a second, take the above

```mlir
#layoutv2_1 = #iree_vector_ext.layoutv2<
  subgroups_per_workgroup = [1, 1],
  batches_per_subgroup = [2, 4],
  duplicates_per_batch = [1, 1],
  threads_per_duplicate = [4, 2],
  elements_per_thread = [1, 4],

  subgroup_order = [0, 1],
  batch_order = [0, 1],
  duplicate_order = [0, 1],
  thread_order = [0, 1],
  element_order = [0, 1]>
```

This takes a vector of `8x32` and, if I'm reading it correctly, is equivalent to

```mlir
#per_dim_1 = #per_dim_layout<[BATCHY, LANEY, VECTORY], [2, 4, 1]>
#per_dim_2 = #per_dim_layout<[BATCHX, LANEX, VECTORX], [4, 2, 4]>
#layout_attr = #layout_attr<#per_dim_1, #per_dim_2>
```

And the element orders in the above give us the freedom to swap X/Y around freely between dimensions

```mlir
#per_dim_1 = #per_dim_layout<[BATCHX, LANEY, VECTORX], [2, 4, 1]>
#per_dim_2 = #per_dim_layout<[BATCHY, LANEX, VECTORY], [4, 2, 4]>
#layout_attr = #layout_attr<#per_dim_1, #per_dim_2>
```

However I don't see a way to swap dimension types around within a dimension using this representation because each level must implicitly tile along the inner or outer portion of each dimension.

```mlir
#per_dim_1 = #per_dim_layout<[VECTORY, LANEY, BATCHY], [1, 4, 2]>
#per_dim_2 = #per_dim_layout<[LANEX, BATCHX, VECTORX], [2, 4, 4]>
#layout_attr = #layout_attr<#per_dim_1, #per_dim_2>
```

Reply:
The original layout, while supporting these dimension swapping, always interpreted them as being ordered in a certain way, even if they weren't. It was always interpreted as BATCH, LANE, VECTOR.

Reply:
Right. The ordering of the labels never meant anything. The thing is that VECTORY, VECTORZ are completely different from VECTORX. I would read VECTORY as DUPLICATEX, and VECTORZ as DUPLICATEY.

Reply:
I think you are thinking of VECTORX, VECTORY as vector dimensions, which is causing the mismatch.

Reply:
Ok now I'm more confused. The example describing the duplication dimensions seems to imply that some threads hold the same elements as other threads (i.e. data is duplicated) but if we look at the `32x32x8` intrinsic, it is doing the following contraction
```
vector.contract ... %a, %b, %c : vector<32x8xf16>, vector<8x32xf16> into vector<32x32xf32>
```
with a subgroup size of 64. The final mfma instruction it produces is
```
amdgpu.mfma %[[AV]] * %[[BV]] + %[[CV]] {blocks = 1 : i32, k = 8 : i32, m = 32 : i32, n = 32 : i32} blgp =  none : vector<4xf16>, vector<4xf16>, vector<16xf32>
```
which has
```
16 elements per thread * 64 threads per subgroup == 1024 == 32 * 32 elements for the original contraction shape
```
so the total number of elements held across all threads seems to match the number of unique elements in the D matrix, so I'm not sure how data is duplicated for the D matrix. cc @harsh-nod because he might have more context on the mfma layouts.

Reply:
Ok maybe i'm just not good with naming things. By duplication I just mean that a thread is holding multiple vectors.

Reply:
0 1
2 3

By duplication i just mean that the layout is duplicated. Not the actual elements.

If you have a shape 8x2:

[x x]
[x x]
[x x]
[x x]

And you say that you are "duplicating" row by 2, it just means that the layout is duplicated.

[0 1]
[2 3]
-------- duplicated layout
[0 1]
[2 3]

Reply:
Ok I would call it `outer_vector`. This really just looks like a detail of the mfma instructions at this point that we have to take in to account because we're lowering to the intrinsic at the same time.